### PR TITLE
Add GapDetectorAgent, an Agent that alerts when no data has been received

### PR DESCRIPTION
**Pull request by cantino**
Similar to PeakDetectorAgent, but this one looks for holes in the data stream.

### CODE REVIEW

**Comment by cantino** [comment]
@dsander, mind a quick review?

File: app/models/agents/gap_detector_agent.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+          {
+            "message": "No data has been received!",
+            "gap_started_at": "1234567890"
+          }
+    MD
+
+    def validate_options
+      unless options['message'].present?
+        errors.add(:base, "window_duration_in_days and message are required")
```
Comment (dsander) [inline app/models/agents/gap_detector_agent.rb]: Isn't that validation just checking the `message` presence?

Comment (cantino) [inline app/models/agents/gap_detector_agent.rb]: Oops, yes, thanks!

File: app/models/agents/gap_detector_agent.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          if event.created_at.to_i > memory['newest_event_created_at']
+            memory['newest_event_created_at'] = event.created_at.to_i
+            memory.delete('alerted_at')
+          end
+        end
+      end
+    end
+
+    def check
+      window = interpolated['window_duration_in_days'].to_f.days.ago
```
Comment (dsander) [inline app/models/agents/gap_detector_agent.rb]: Neat, I did not know `.days` works with floats.

File: app/models/agents/gap_detector_agent.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      if memory['newest_event_created_at'].present? && Time.at(memory['newest_event_created_at']) < window
```
Comment (dsander) [inline app/models/agents/gap_detector_agent.rb]: Maybe add `self.created_at > window` to the condition?

Comment (cantino) [inline app/models/agents/gap_detector_agent.rb]: Would that be to fix initial state issues?  I thought that `memory['newest_event_created_at'].present?` would prevent issues with the startup state, but I may be missing something.

Comment (dsander) [inline app/models/agents/gap_detector_agent.rb]: Oh, you are right. I missed that.

**Comment by dsander** [comment]
Looks good to me :+1:

**Comment by cantino** [comment]
Thanks!

**Comment by cantino** [comment]
Will merge once I see it all work locally. :smile:

**Pull request merged**